xapian-applications/omega/clickmodel/tests/sdbntest.cc

```diff
@@ -0,0 +1,130 @@
+/** @file sdbntest.cc
+ * @brief tests for the Simplified DBN clickmodel class.
+ */
+/* Copyright (C) 2017 Vivek Pal
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+
+#include <cmath>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include <clickmodel/simplifieddbn.h>
+
+using namespace std;
+
+#define QID 0
+#define DOCIDS 1
+#define CLICKS 2
+
+struct sessions_items {
+    const string qid;
+    const string docids;
+    const string clicks;
+};
+
+struct sessions_testcase {
+    const string logfile;
+    const sessions_items sessions;
+};
+
+string sample_log1 = "test1.log";
+string sample_log2 = "test2.log";
+string sample_log3 = "test3.log";
+
+static sessions_testcase sessions_tests[] = {
+    {sample_log1, {"821f03288846297c2cf43c34766a38f7",
+		   "45,36,14,54,42",
+		   "45:0,36:0,14:0,54:2,42:0"}},
+    {sample_log2,{}}
+};
+
+static bool
+almost_equal(const double &x, const double &y,
+			 const double epsilon = 0.001) {
+    return (abs(x - y) < epsilon) ? true : false;
+}
+
+int main() {
+    SimplifiedDBN sdbn;
+
+    // Tests for SimplifiedDBN::build_sessions method.
+    for (size_t i = 0; i < sizeof(sessions_tests) / sizeof(sessions_tests[0]); ++i) {
+	vector<vector<string>> result;
+	try {
+	    result = sdbn.build_sessions(sessions_tests[i].logfile);
+	} catch (std::exception &ex) {
+	    cout << ex.what() << endl;
+	    exit(1);
```

Comment:
Rather than exit in this loop, you should probably note that a failure has occurred and keep going, so all tests are run. (Otherwise if you break things you'll fix the first failure, but not know about the second until you run the tests again — which is a little annoying. With a small number of tests it's not bad, but if this gets more complex it could slow things down a lot.)